exodus_lambda/functions/origin_request.py

```diff
@@ -101,6 +100,10 @@ def handler(self, event, context):
                 request["uri"] = (
                     "/" + query_result["Items"][0]["object_key"]["S"]
                 )
+                if query_result["Items"][0]["content_type"]["S"]:
+                    request["uri"] += "&ResponseContentType={}".format(
```

Comment:
Content type should be separate from uri, i.e., `request["content_type"] = query_result["Items"][0]["content_type"]["S"]`.

Reply:
The comment about request["content_type"] doesn't seem right to me.

The goal here is to rewrite the request so that, if the DB says the response should be content-type X, then we arrange for S3 to respond with content-type X.  It can be achieved by including ResponseContentType in the query string before it goes to S3, documented at https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#API_GetObject_RequestSyntax (note: a little unclear from the docs if it's "response-content-type" or "ResponseContentType", guess it's the latter)

Setting request["content_type"] seems like it'd do something else... like overriding the content type of the incoming request from the user?  Or maybe it's invalid, I couldn't see any reference to content_type outside of headers at https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-event-structure.html#lambda-event-structure-request .

I think the overall approach here is right but the implementation seems slightly off:

- Adding & is only valid if there's already a non-empty query string, right?  Otherwise it should be ? to start the query string?
- the value put into the query string should be url-encoded.

Ideally we're using some library function to build the query string rather than doing it ad-hoc.

Reply:
About the URL encoding, what I meant is something like this:

```
                    content_type = query_result["Items"][0]["content_type"]["S"]
                    query = urllib.parse.urlencode({'ResponseContentType': content_type})
                    request["uri"] += "?" + query
```

If content_type contains any special characters, they'll be encoded so that the other end can receive them properly, e.g. text/plain becomes text%2Fplain.

When I made the suggestion originally, I thought that it might be important to encode the `/` character.  But some of the APIs in urllib call it "safe", so now I'm not so sure.  That means I'm not 100% sure that any content types actually used will contain any characters which have to be encoded, but it's probably best to be on the safe side and encode anyway.

Reply:
Ah, I clearly didn't know what's going on. Thanks for the doc link, @rohanpm.

Reply:
The AWS documentation is confusing.  Let me double check.

Reply:
That is pretty strange, but some examples show there isn't a literal `+`, e.g. there's an example like `GET /Junk3.txt?response-cache-control=No-cache`.

I have a feeling the `+` is meant to communicate something else, maybe the fact that `Key` is always at least one character.

The example now makes me question whether it should be ResponseContentType or response-content-type though (maybe both work?)

Reply:
@rohanpm Is correct. The `+` is not required. Double checked that. In regards to the ResponseContentType vs. response-content-type I don't know the answer yet. Let's defer that to a later stage.